### PR TITLE
Adding utility to register a custom element

### PR DESCRIPTION
**Pull request by rorticus**
**Type:** feature

The following has been addressed in the PR:

* [ ] There is a related issue
* [x] All code matches the [style guide](https://github.com/dojo/meta/blob/master/STYLE.md)
* [x] Unit or Functional tests are included in the PR

**Description:**

This is another chunk of custom elements code. The code in this PR will actually register a custom element with the browser. The intent is that the actual method call will come from generated code that the tooling.

There are no tests for this as I don't really know how to test this since it's doing things specific to DOM element creation...

### CODE REVIEW

File: src/registerCustomElement.ts [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+		}
+	});
+}
+
+/**
+ * Register a custom element using the v0 spec of custom elements.
+ *
+ * @param descriptorFactory
+ */
+export function registerCustomElementV0(descriptorFactory: CustomElementDescriptorFactory) {
```
Comment (matt-gadd) [inline src/registerCustomElement.ts]: I think we should ditch the V0 support, as far as i'm aware only chrome implemented it and so it doesn't give us any more support.

Comment (rorticus) [inline src/registerCustomElement.ts]: Can do. technically it works in IE 10 while the V1 support doesn't. But if we're not interested in that, it's outta here!

Comment (rorticus) [inline src/registerCustomElement.ts]: Crushed v0 support with an iron fist.

File: src/registerCustomElement.ts [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+}
+
+/**
+ * Register a custom element using the v1 spec of custom elements. Note that
+ * this is the default export, and, expects the proposal to work in the browser.
+ * This will likely require the polyfill and native shim.
+ *
+ * @param descriptorFactory
+ */
+export function registerCustomElementV1(descriptorFactory: CustomElementDescriptorFactory) {
```
Comment (agubler) [inline src/registerCustomElement.ts]: Do we need the `V1` suffix?

Comment (rorticus) [inline src/registerCustomElement.ts]: Good point! I guess we don't need that anymore...

File: src/registerCustomElement.ts [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+		getWidgetFactory(): WidgetFactory<any, any> {
+			return this.getDescriptor().widgetFactory;
+		}
+
+		getDescriptor(): CustomElementDescriptor {
+			return descriptor;
+		}
+
+		static get observedAttributes(): string[] {
+			return (descriptor.attributes || []).map(attribute => attribute.attributeName);
```
Comment (agubler) [inline src/registerCustomElement.ts]: there is a missing test branch here.

File: src/customElements.ts [hunk 4]
```diff
@@ -129,6 +131,23 @@ function getWidgetPropertyFromAttribute(attributeName: string, attributeValue: s
 	return [ propertyName, value ];
 }
 
+let customEventClass = global.CustomEvent;
+
+if (typeof customEventClass !== 'function') {
+	const customEvent = function (event: string, params: any) {
+		params = params || { bubbles: false, cancelable: false, detail: undefined };
+		let evt = document.createEvent('CustomEvent');
```
Comment (agubler) [inline src/customElements.ts]: can this be a `const`

File: src/registerCustomElement.ts (hunk 3 above)
Comment (rorticus) [inline src/registerCustomElement.ts]: picky picky :)

File: src/customElements.ts (hunk 4 above)
Comment (rorticus) [inline src/customElements.ts]: yup

File: src/registerCustomElement.ts (hunk 3 above)
Comment (agubler) [inline src/registerCustomElement.ts]: That's the difference between codecov checks passing and failing 😄

Comment (rorticus) [inline src/registerCustomElement.ts]: Haha! Zing!

**Pull request merged**